design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,346 @@
+xlang_pointer              | 0x80000005 | Null pointer related errors
+xlang_not_impl             | 0x80000001 | Function / feature is not implemented
+xlang_access_denied        | 0x80070005 | Requested operation / call is not allowed
+xlang_no_interface         | 0x80000004 | Class or interface not found
+xlang_handle               | 0x80000006 | Invalid handle
+xlang_abort                | 0x80000007 | Operation / call has been canceled
+xlang_file_not_found       | 0x80070002 | File was not found
+xlang_path_not_found       | 0x80070003 | Directory was not found
+xlang_arithmetic           | 0x80070216 | Error during arithmetic operation
+xlang_io                   | 0x80131620 | IO error than does not have a better error mapping
+xlang_bounds               | 0x8000000b | Invalid index used such as when indexing arrays
+xlang_timeout              | 0x80131505 | Operation / call timed out
+xlang_thread_state         | 0x80131520 | Operation can not be done due to the thread state
+xlang_thread_interrupted   | 0x80131519 | Interrupt on thread
+xlang_security             | 0x8013150A | Security related error
+xlang_type_load            | 0x80131522 | Syntax / type error during runtime
+
+Even though xlang only defines these common errors, it will not prevent the use of other errors from
+the HRESULT family. But, it is likely that those errors will not get a natural mapping from
+projections and end up as generic errors.
+
+## How to indicate the result of a function
+
+Functions declared at the xlang ABI (i.e. language projections, PAL) need a way to indicate whether
+an error occurred. This will be done through the return parameter.
+
+```
+xlang_result functionName(param, ...)
+```
+
+Returning one of the error codes will indicate that the function resulted in an error.
+Returning xlang_ok which has a value of 0 will indicate the function completed without errors.
+
+## Alternative to errors
+
+Even though xlang does provide a set of errors to assist with the error translation from one language
+to another in an equivalent form, it is limited. In addition, when errors are translated from one language
+to another, there is a chance that the accuracy of the error is lost due to an exact error mapping is not
+available in the other language. This means the caller of an API may not always have the accurate and
+necessary information to decide what to do with a failure.
+
+Due to this, xlang encourages the use of errors only for diagnosing failures and not for controlling
+code flow. For controlling code flow, xlang encourages the use of the Try pattern in functions.
+The Try pattern can be used to pass back custom domain specific errors as parameters if necessary.
+The parameter type and its possible values are declared in the producing component and would be translated
+across languages as is. This means the consumer can rely on using it to make decisions without being
+concerned of the translation accuracy.
+
+## Diagnosing application failures
```

Comment:
Mike brought up the idea of stowed exceptions - when an exception gets ignored but then end the code ends up in a bad case down the line and raises a new exception. Is there a way to get access to the details of the ignored exception?

Reply:
Kenny called this "proximity debugging"

Reply:
As per arch meeting, we want to consider keeping a buffer of errors, but that such a feature is outside the scope of this DN